### PR TITLE
 Revert "`RingBufferLogHandler.records` could leak records

### PR DESCRIPTION
**Pull request by timja**
See [JENKINS-XXXXX](https://issues.jenkins-ci.org/browse/JENKINS-XXXXX).

See https://github.com/jenkinsci/jenkins/pull/6018#issuecomment-990367907

This change isn't released yet so should get it in ASAP (unless someone can see the issue in the original PR, I don't have time to debug it though)

#6018 broke the controller System log.

Picked up in ATH as it uses the log to check plugins have completed installing.

I wasn't able to reproduce this with `JenkinsRule` but trivially reproduced with `RealJenkinsRule` added a test in https://github.com/jenkinsci/jenkins/commit/da1cc333182eabc430d8277ab0cd6e396f8b24b4 which failed on master and passes after reverting

### Proposed changelog entries

* Entry 1: Issue, Human-readable Text
* ...

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by MarkEWaite — APPROVED** [review]

File: core/src/test/java/hudson/logging/LogRecorderTest.java [hunk 1]
```diff
@@ -115,7 +113,12 @@
         lr.handler.publish(r5);
         lr.handler.publish(r6);
 
-        assertThat(lr.handler.getView(), contains(r6, r5, r1));
+        assertTrue(lr.handler.getView().contains(r1));
+        assertFalse(lr.handler.getView().contains(r2));
+        assertFalse(lr.handler.getView().contains(r3));
+        assertFalse(lr.handler.getView().contains(r4));
+        assertTrue(lr.handler.getView().contains(r5));
+        assertTrue(lr.handler.getView().contains(r6));
```
Comment (jglick) [inline core/src/test/java/hudson/logging/LogRecorderTest.java:121]: Certainly do not revert this file!

File: test/src/test/java/hudson/logging/LogRecordManagerRealTest.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    public void logRecordsArePresentOnController() throws Throwable {
+        rr.then(new LogRecordsArePresent());
+    }
+
+    private static class LogRecordsArePresent implements RealJenkinsRule.Step {
+
+        @Override
+        public void run(JenkinsRule r) throws Throwable {
+            List<LogRecord> logRecords = Jenkins.logRecords;
+            assertThat(logRecords, is(not(empty())));
```
Comment (jglick) [inline test/src/test/java/hudson/logging/LogRecordManagerRealTest.java:31]: Let me take a quick look using this test to see if I can spot the problem in the original.

File: test/src/test/java/hudson/logging/LogRecordManagerRealTest.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+    @Rule
+    public RealJenkinsRule rr = new RealJenkinsRule();
+
+    @Test
+    public void logRecordsArePresentOnController() throws Throwable {
+        rr.then(new LogRecordsArePresent());
+    }
+
+    private static class LogRecordsArePresent implements RealJenkinsRule.Step {
```
Comment (jglick) [inline test/src/test/java/hudson/logging/LogRecordManagerRealTest.java:26]: FYI this is conventionally done via a `static` method and a method reference. See Javadoc.

File: test/src/test/java/hudson/logging/LogRecordManagerRealTest.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+import java.util.List;
+import java.util.logging.LogRecord;
+import jenkins.model.Jenkins;
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.jvnet.hudson.test.RealJenkinsRule;
+
+import static org.hamcrest.MatcherAssert.assertThat;
```
Comment (jglick) [inline test/src/test/java/hudson/logging/LogRecordManagerRealTest.java:11]: Checkstyle complaining

**Comment by jglick** [comment]
Ah, I get it: https://github.com/jenkinsci/jenkins/blob/da1cc333182eabc430d8277ab0cd6e396f8b24b4/core/src/main/java/hudson/WebAppMain.java#L330 is the reason for the weird `getView` implementation. Foolish to have this be a `public` field to begin with, but we are stuck with that.

**Comment by jglick** [comment]
Let me try to file a separate PR using your test and just fixing `getView`.

File: test/src/test/java/hudson/logging/LogRecordManagerRealTest.java (hunk 4 above)
Comment (timja) [reply test/src/test/java/hudson/logging/LogRecordManagerRealTest.java:11]: aware, just left it while you were debugging

File: test/src/test/java/hudson/logging/LogRecordManagerRealTest.java (hunk 3 above)
Comment (timja) [reply test/src/test/java/hudson/logging/LogRecordManagerRealTest.java:26]: ah cool feel free to adjust, for some reason by IDE wouldn't download javadoc while I was writing it so I just searched the org for examples 😢

**Comment by jglick** [comment]
#6044 should do it I think.

**Pull request closed without merging**